test/hotspot/gtest/riscv/test_assembler_riscv.cpp

```diff
@@ -341,132 +436,293 @@ TEST_VM(RiscV, cmpxchg_int8_maybe_zacas) {
   }
 }
 
+template <typename TESTSIZE, Assembler::operand_size ASMSIZE>
+static void run_concurrent_narrow_cmpxchg_tests() {
+  volatile TESTSIZE data = 0;
+  int num_threads = 4;
+  NarrowCmpxchgTester<TESTSIZE, ASMSIZE> cmpxchg(false); // not bool ret
+  auto incThread = [&](Thread* _current, int _id) {
+    for (int i = 0; i < 10000; i++) {
+      TESTSIZE oldvalue = _id + num_threads * i;
+      TESTSIZE newvalue = oldvalue + 1;
+      TESTSIZE ret;
+      do {
+        ret = cmpxchg.narrow_cmpxchg((intptr_t)&data, oldvalue, newvalue);
+      } while (ret != oldvalue);
+    }
+  };
+  TestThreadGroup<decltype(incThread)> ttg(incThread, num_threads);
+  ttg.doit();
+  ttg.join();
+  ASSERT_EQ(data, (TESTSIZE)(num_threads*10000));
+}
+
+template <typename TESTSIZE, Assembler::operand_size ASMSIZE>
+static void run_concurrent_alt_narrow_cmpxchg_tests() {
+  volatile TESTSIZE data = 0;
+  int num_threads = 4;
+  NarrowCmpxchgTester<TESTSIZE, ASMSIZE> cmpxchg(false); // not bool ret
+  auto incThread = [&](Thread* _current, int _id) {
+    for (int i = 0; i < 10000; i++) {
+      TESTSIZE oldvalue;
+      TESTSIZE ret = 0;
+      do {
+        oldvalue = ret;
+        TESTSIZE newvalue = oldvalue + 1;
+        ret = cmpxchg.narrow_cmpxchg((intptr_t)&data, oldvalue, newvalue);
+      } while (ret != oldvalue);
+    }
+  };
+  TestThreadGroup<decltype(incThread)> ttg(incThread, num_threads);
+  ttg.doit();
+  ttg.join();
+  ASSERT_EQ(data, (TESTSIZE)(num_threads*10000));
```

Comment:
Strictly speaking the first overflow happens here:
`TESTSIZE newvalue = oldvalue + 1;`
As '1' is an integer, oldvalue is promoted to an integer. As 128 is not in range of int8_t.
This is implementation defined, but de facto we know this will wrap around.

The casting of 40000 to TESTIZE the same thing happens, but first the higher bits are truncated and then it may wrap around depending on the MSB.

AFIACT it always produces the same result.

If we don't trust the wrap around as it's outside the spec both cases should be handled.

Reply:
Yes, but doing just 127 incs there, then it is big chance of the threads not running at the same time.
The point is that they should run parallel a period of time.

Reply:
Seems to me overflow is fine, it won't affect the test correctness? maybe we could add some comment to state it clearly.
Or maybe we could use something like barrier or phaser in java API to sync the real starting of different threads in the thread group? But I'm not sure if there are such utilities in hotspot test framework.

Reply:
> So maybe: Run loops from TESTSIZE min to max, re-run this loop until we record CAS failures or something like that?

Seens to me that would be very similar with the current approach. How about we make the wrap around here explicit? Say, when the `oldvalue` reaches TESTSIZE max, we set the `newvalue` to TESTSIZE min. And we set the final expect value in `data` accordingly instead of doing a type truncation. Then we don't depend on the implementation of the compiler.

Reply:
The truncation is well defined, it's just signess overflow that is not spec.
As TESTSIZE may be 64 bits and we don't simply increase by one that check itself will overflow.

Note that even if compiler do something else than wrap around, it would do the same in both cases so it should produce the same result.

Also note that this cast is no worse than doing (signed char)0xff.

Anyhow I tested trying to add max and min, the code just becomes very ugly and complicated.

Reply:
> The truncation is well defined, it's just signess overflow that is not spec. As TESTSIZE may be 64 bits and we don't simply increase by one that check itself will overflow.

Ah, I was considering TESTSIZE int8_t or int16_t.

> Note that even if compiler do something else than wrap around, it would do the same in both cases so it should produce the same result.

That makes sense to me. Thanks.

> Also note that this cast is no worse than doing (signed char)0xff.
> 
> Anyhow I tested trying to add max and min, the code just becomes very ugly and complicated.

All right then. Could you please add some code comment about what we are doing here to help understand?

Reply:
I tried to fix it, let me know.